web/src/main/java/org/springframework/security/web/csrf/XorCsrfTokenRepository.java

```diff
@@ -0,0 +1,198 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.web.csrf;
+
+import java.security.MessageDigest;
+import java.security.SecureRandom;
+import java.util.Base64;
+
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+
+import org.springframework.security.crypto.codec.Utf8;
+import org.springframework.util.Assert;
+
+/**
+ * A {@link CsrfTokenRepository} that wraps another implementation with the ability to
+ * provide {@link CsrfToken} objects that are salted with a random value on every request.
+ *
+ * @author Steve Riesenberg
+ */
+public final class XorCsrfTokenRepository implements CsrfTokenRepository {
+
+	private final CsrfTokenRepository delegate;
+
+	private final SecureRandom secureRandom;
+
+	/**
+	 * Creates a new instance of {@link XorCsrfTokenRepository}.
+	 * @param delegate The {@link CsrfTokenRepository} to delegate to for persistence
+	 */
+	public XorCsrfTokenRepository(CsrfTokenRepository delegate) {
+		this(delegate, new SecureRandom());
+	}
+
+	/**
+	 * Creates a new instance of {@link XorCsrfTokenRepository}.
+	 * @param delegate The {@link CsrfTokenRepository} to delegate to for persistence
+	 * @param secureRandom The {@link SecureRandom} to use for generating salt values
+	 */
+	public XorCsrfTokenRepository(CsrfTokenRepository delegate, SecureRandom secureRandom) {
+		Assert.notNull(delegate, "delegate cannot be null");
+		Assert.notNull(secureRandom, "secureRandom cannot be null");
+		this.delegate = delegate;
+		this.secureRandom = secureRandom;
+	}
+
+	@Override
+	public CsrfToken generateToken(HttpServletRequest request) {
+		CsrfToken csrfToken = this.delegate.generateToken(request);
+		return new XorCsrfToken(csrfToken, this.secureRandom);
+	}
+
+	@Override
+	public void saveToken(CsrfToken token, HttpServletRequest request, HttpServletResponse response) {
+		CsrfToken tokenToSave = token;
+		if (token instanceof XorCsrfToken) {
+			tokenToSave = ((XorCsrfToken) token).getCsrfToken();
+		}
+		this.delegate.saveToken(tokenToSave, request, response);
+	}
+
+	@Override
+	public CsrfToken loadToken(HttpServletRequest request) {
+		CsrfToken csrfToken = this.delegate.loadToken(request);
+		return new XorCsrfToken(csrfToken, this.secureRandom);
+	}
+
+	/**
+	 * Wrap an underlying {@link CsrfTokenRepository} with the ability to provide
+	 * {@link CsrfToken} objects that are salted with a random value on every request.
+	 * @param delegate The {@link CsrfTokenRepository} to delegate to for persistence
+	 * @return An {@link XorCsrfTokenRepository} that wraps the given
+	 * {@link CsrfTokenRepository}
+	 */
+	public static CsrfTokenRepository wrap(CsrfTokenRepository delegate) {
```

Comment:
Note: I'm unsure whether we want a static method or just use the constructor since historically other implementations have just had a `public` constructor.

Reply:
I'd just keep the constructor. Static methods are useful when you have lots of constructors and need more descriptive ways of creating an instance. It is also useful when the constructor can take the same number of arguments/types but the arguments are different. For example, `static ABC createFromToken(String)` and `static AbC createFromUrl(String)` are necessary because the number of arguments and types of the arguments are the same, but the value provided is different meaning.